### ISSUE TITLE
Dropping implementation of ArrayStorage

### ISSUE THREAD

**Issue by Nucs** [SciSharp/NumSharp#294]
After a deep look at the backend architecture and doing some experiments, I would like you to consider the following benchmark:

![image](https://user-images.githubusercontent.com/649919/60176860-27770e00-9820-11e9-928e-4f8e068df9d0.png)

```C#
public double[] genericArray;
public Array nonGenericArray;

[GlobalSetup]
public void Setup()
{
    var rnd = new Random(42);
    // first array
    nonGenericArray = genericArray = new double[10_000_000];

    for (int i = 0; i < genericArray.Length; i++)
    {
        genericArray[i] = rnd.NextDouble();
    }
}

[Benchmark(Baseline = true)]
public void GenericAccess()
{
    var length = genericArray.Length;
    for (int i = 0; i < length; i++)
    {
        var a = genericArray[i];
    }
}

[Benchmark]
public void NonGenericAccess()
{
    double sum = 0;
    var length = nonGenericArray.Length;
    for (int i = 0; i < length; i++)
    {
        var a = nonGenericArray.GetValue(i);
    }
}
```
### Proposition
I propose to drop-off non-generic `ArrayStorage` and only use `TypedGenericStorage`.

I have already done it in my local but I can't remove base classes without other architects's review.

**Comment by Oceania2018** [comment]
@Nucs Could you push `GenericAccess` into repo ? I can't find it.

![image](https://user-images.githubusercontent.com/1705364/60179846-3a6bed00-97e4-11e9-8f24-377f580a1fec.png)

**Comment by henon** [comment]
My 2 cents: 
* if the benchmarks are in favor of your proposal (wow a factor 100 in some cases!!!)
* all tests work with the new storage
Why the hell not ;)

This is a huge change though. Can you pull it off?

**Comment by Nucs** [comment]
@henon, Already did, looking for approval.
I can't PR yet but I can't continue without the approval
@Oceania2018 Added.

**Comment by Oceania2018** [comment]
Just push to your repo.

**Comment by henon** [comment]
Awesome @Nucs, I am looking forward to the PR.

**Comment by Oceania2018** [comment]
@Nucs Thank you for your nice comparison, actually, We're using the `TypedArrayStorage` not `ArrayStorage`, I've already found the big performance difference before.

I refactored it in `TypedArrayStorage`.

![image](https://user-images.githubusercontent.com/1705364/60227521-4ba11200-9855-11e9-85e2-d2d2afb91e1b.png)

Our long term goal is implement another `Storage` -> `ByteStorage` to implement the way of zero-copy,
and share data with other libraries like `ArrayFire`, `Apache Arrow`.
